### DISCUSSION TITLE
How can I delete query key from query params ?

### DISCUSSION THREAD

**Discussion by bishalDores** [vercel/next.js#53555] [App Router] [answered]
Hi there.  I am filtering data in a page through query params. I am using checkboxes to update params in the url. Upon checked params are added in the query and while unchecking params are being removed. The problem is,  while all the checkboxes are unchecked, the checkbox name which is  the 'query key' is still there. For example:
http://localhost:3000/?jobType=Permanent--Internship
Here 'jobType' is the key. While Permanent and Internship are removed from the params, I am deleting the key 'jobType' and pushing the route to "/", but '?jobType' is still there in the url like: http://localhost:3000/?jobType= ? 
How can I remove '?jobType='  from the url also? Below is the code.

"use client";

import React, { useState, useEffect } from "react";
import { useRouter, usePathname, useSearchParams, useParams } from "next/navigation";

const FilterOne = () => {
  const [checkboxValues, setCheckboxValues] = useState([]);
  const router = useRouter();
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const params = new URLSearchParams(searchParams);

  const onChangeHandler = (e) => {
    const { name, value, checked } = e.target;
    const newValues = checked ? [...checkboxValues, value] : checkboxValues.filter((val) => val !== value);

    setCheckboxValues(newValues);

    if (newValues.length === 0) {
      params.forEach((value, key) => {
        params.delete(key);
      });

      console.log("params", params);
      router.push(pathname);
    }
    if (newValues.length > 1) {
      params.set(name, newValues.join("--"));
      router.push(pathname + "?" + params.toString());
    } else {
      params.set(name, newValues.join(""));
      router.push(pathname + "?" + params.toString());
    }
  };
  return (
    <div className="sidebar mt-5">
      <h3>Filters</h3>

      <hr />
      <h5 className="filter-heading mb-3">Job Type</h5>

      <div className="form-check">
        <input
          className="form-check-input"
          type="checkbox"
          name="jobType"
          id="check1"
          value="Permanent"
          onChange={onChangeHandler}
          checked={checkboxValues.includes("Permanent")}
        />
        <label className="form-check-label" htmlFor="check1">
          Permanent
        </label>
      </div>

      <div className="form-check">
        <input
          className="form-check-input"
          type="checkbox"
          name="jobType"
          id="check2"
          value="Temporary"
          onChange={onChangeHandler}
          checked={checkboxValues.includes("Temporary")}
        />
        <label className="form-check-label" htmlFor="check2">
          Temporary
        </label>
      </div>

      <div className="form-check">
        <input
          className="form-check-input"
          type="checkbox"
          name="jobType"
          id="check3"
          value="Internship"
          onChange={onChangeHandler}
          checked={checkboxValues.includes("Internship")}
        />
        <label className="form-check-label" htmlFor="check3">
          Internship
        </label>
      </div>
    </div>
  );
};

export default FilterOne;

**Comment by jeandv** [comment]
in your code, you are only deleting the key from the params object and not updating the URL. You can try updating the URL by calling router.push(pathname) after deleting the key from the params object. This should remove the jobType key from the URL.

```
if (newValues.length === 0) {
  params.forEach((value, key) => {
    params.delete(key);
  });

  console.log("params", params);
  router.push(pathname);
}
```

**Reply by bishalDores** [reply]
I am already updating the path after deleting the key as mentioned in the above code but its not working. Are you suggesting something else ? maybe I am not getting you.  Can you provide the code please ?

**Comment by Nadeera3784** [comment, marked as answer]
Hi @bishalDores  maybe you can try something like this. 

```
const onChangeHandler = (e) => {
  const { name, value, checked } = e.target;
  const newValues = checked ? [...checkboxValues, value] : checkboxValues.filter((val) => val !== value);

  setCheckboxValues(newValues);

  if (newValues.length === 0) {
    params.delete(name); // Remove the entire 'jobType' key from the params
  } else {
    params.set(name, newValues.join("--")); // Set the new values for 'jobType' key
  }

  router.push(pathname + (params.toString() ? "?" + params.toString() : ""));
};
```
when all checkboxes are unchecked, the 'jobType' key will be removed from the URL query, and the URL will look like http://localhost:3000/ without any query parameters.

**Reply by bishalDores** [reply]
Thanks @Nadeera3784
